File: ts/src/lib/range-tree.ts
import { RangeCov } from "./types";

export class RangeTree {
  start: number;
  end: number;
  count: number;
  children: RangeTree[];
  parentIndex: number;
  private lazyCount: number;

  constructor(
    start: number,
    end: number,
    count: number,
    children: RangeTree[],
    lazyCount: number,
    parentIndex: number,
  ) {
    this.start = start;
    this.end = end;
    this.count = count;
    this.children = children;
    this.lazyCount = lazyCount;
    this.parentIndex = parentIndex;
  }

  /**
   * @precodition `ranges` are well-formed and pre-order sorted
   */
  static fromSortedRanges(ranges: ReadonlyArray<RangeCov>): RangeTree | undefined {
    let root: RangeTree | undefined;
    const stack: RangeTree[] = [];
    for (const range of ranges) {
      const node: RangeTree = new RangeTree(range.startOffset, range.endOffset, range.count, [], 0, 0);
      if (root === undefined) {
        root = node;
        stack.push(node);
        continue;
      }
      let top: RangeTree;
      while (true) {
        // assert: `top.length > 0`
        top = stack[stack.length - 1];
        if (range.startOffset >= top.end) {
          stack.pop();
          top = stack[stack.length - 1];
        } else {
          break;
        }
      }
      top.children.push(node);
      stack.push(node);
    }
    return root;
  }

  normalize(): void {
    if (this.lazyCount !== 0) {
      this.count += this.lazyCount;
      for (const child of this.children) {
        child.lazyCount += this.lazyCount;
      }
      this.lazyCount = 0;
    }

    const children: RangeTree[] = [];
    let curEnd: number;
    let head: RangeTree | undefined;
    const tail: RangeTree[] = [];
    for (const child of this.children) {
      if (head === undefined) {
        head = child;
      } else if ((child.count + child.lazyCount) === (head.count + head.lazyCount) && child.start === curEnd!) {
        tail.push(child);
      } else {
        endChain();
        head = child;
      }
      curEnd = child.end;
    }
    if (head !== undefined) {
      endChain();
    }

    if (children.length === 1) {
      const child: RangeTree = children[0];
      if (child.start === this.start && child.end === this.end) {
        this.count = child.count;
        this.children = child.children;
        // `.lazyCount` is zero for both (both are after normalization)
        // `.parentIndex` is irrelevant
        return;
      }
    }

    this.children = children;

    function endChain(): void {
      if (tail.length !== 0) {
        head!.end = tail[tail.length - 1].end;
        for (const tailTree of tail) {
          for (const subChild of tailTree.children) {
            subChild.lazyCount += tailTree.lazyCount - head!.lazyCount;
            head!.children.push(subChild);
          }
        }
        tail.length = 0;
      }
      head!.normalize();
      children.push(head!);
    }
  }

  /**
   * @precondition `tree.start < value && value < tree.end`
   * @return RangeTree Right part
   */
  split(value: number): RangeTree {
    let leftEnd: number = this.children.length;
    let mid: RangeTree | undefined;

    // TODO(perf): Binary search (check overhead)
    for (let i: number = 0; i < this.children.length; i++) {
      const child: RangeTree = this.children[i];
      if (child.start < value && value < child.end) {
        mid = child.split(value);
        leftEnd = i + 1;
        break;
      } else if (child.start >= value) {
        leftEnd = i;
        break;
      }
    }

    const rightLen: number = this.children.length - leftEnd;
    const rightChildren: RangeTree[] = this.children.splice(leftEnd, rightLen);
    if (mid !== undefined) {
      rightChildren.unshift(mid);
    }
    const result: RangeTree = new RangeTree(
      value,
      this.end,
      this.count,
      rightChildren,
      this.lazyCount,
      this.parentIndex,
    );
    this.end = value;
    return result;
  }

  /**
   * Get the range coverages corresponding to the tree.
   *
   * The ranges are pre-order sorted.
   */
  toRanges(): RangeCov[] {
    const ranges: RangeCov[] = [];
    const stack: RangeTree[] = [this];
    while (stack.length > 0) {
      const cur: RangeTree = stack.pop()!;
      ranges.push({startOffset: cur.start, endOffset: cur.end, count: cur.count});
      for (let i: number = cur.children.length - 1; i >= 0; i--) {
        stack.push(cur.children[i]);
      }
    }
    return ranges;
  }

  /**
   * Increases the count of the current tree and all its descendants.
   *
   * The tree and its descendants become invalidated: you need to normalize it
   * before reading it.
   *
   * @param n Value to add.
   */
  addCount(n: number): void {
    this.lazyCount += n;
  }
}

File: rs/src/range_tree.rs
use coverage::RangeCov;
use std::iter::Peekable;

#[derive(Eq, PartialEq, Clone, Debug)]
pub struct RangeTree {
  pub start: usize,
  pub end: usize,
  pub count: i64,
  pub children: Vec<RangeTree>,
}

impl RangeTree {
  pub fn new(start: usize, end: usize, count: i64, children: Vec<RangeTree>) -> RangeTree {
    Self {
      start,
      end,
      count,
      children,
    }
  }

  pub fn split(self, value: usize) -> (Self, Self) {
    let mut left_children: Vec<Self> = Vec::new();
    let mut right_children: Vec<Self> = Vec::new();
    for child in self.children {
      if child.end <= value {
        left_children.push(child);
      } else if value <= child.start {
        right_children.push(child);
      } else {
        let (left_child, right_child) = child.split(value);
        left_children.push(left_child);
        right_children.push(right_child);
      }
    }

    let left = RangeTree::new(self.start, value, self.count, left_children);
    let right = RangeTree::new(value, self.end, self.count, right_children);
    (left, right)
  }

  pub fn normalize(mut self) -> Self {
    self.children = {
      let mut children: Vec<RangeTree> = Vec::new();
      let mut chain: Vec<RangeTree> = Vec::new();
      for child in self.children.into_iter() {
        let is_chain_end: bool = match chain.last().map(|tree| (tree.count, tree.end)) {
          Some((count, chain_end)) => (count, chain_end) != (child.count, child.start),
          None => false,
        };
        if is_chain_end {
          let mut chain_iter = chain.drain(..);
          let mut head: RangeTree = chain_iter.next().unwrap();
          for tree in chain_iter {
            head.end = tree.end;
            head.children.extend(tree.children.into_iter());
          }
          children.push(head.normalize());
        }
        chain.push(child)
      }
      if !chain.is_empty() {
        let mut chain_iter = chain.drain(..);
        let mut head: RangeTree = chain_iter.next().unwrap();
        for tree in chain_iter {
          head.end = tree.end;
          head.children.extend(tree.children.into_iter());
        }
        children.push(head.normalize());
      }

      if children.len() == 1 && children[0].start == self.start && children[0].end == self.end {
        let child = children.remove(0);
        self.count = child.count;
        children = child.children;
      }

      children
    };

    self
  }

  pub fn add_count(&mut self, value: i64) -> () {
    self.count += value;
    for child in self.children.iter_mut() {
      child.add_count(value);
    }
  }

  pub fn to_ranges(&self) -> Vec<RangeCov> {
    let mut ranges: Vec<RangeCov> = Vec::new();
    let mut stack: Vec<&RangeTree> = vec![self];
    while let Some(ref cur) = stack.pop() {
      ranges.push(RangeCov { start_offset: cur.start, end_offset: cur.end, count: cur.count });
      for child in cur.children.iter().rev() {
        stack.push(child)
      }
    }
    ranges
  }

  pub fn from_sorted_ranges(ranges: &[RangeCov]) -> Option<RangeTree> {
    Self::from_sorted_ranges_inner(&mut ranges.iter().peekable(), ::std::usize::MAX)
  }

  fn from_sorted_ranges_inner<'a, 'b: 'a>(ranges: &'a mut Peekable<impl Iterator<Item=&'b RangeCov>>, parent_end: usize) -> Option<RangeTree> {
    let has_range: bool = match ranges.peek() {
      None => false,
      Some(ref range) => range.start_offset < parent_end,
    };
    if !has_range {
      return None;
    }
    let range = ranges.next().unwrap();
    let start: usize = range.start_offset;
    let end: usize = range.end_offset;
    let count: i64 = range.count;
    let mut children: Vec<RangeTree> = Vec::new();
    while let Some(child) = Self::from_sorted_ranges_inner(ranges, end) {
      children.push(child);
    }
    Some(RangeTree::new(start, end, count, children))
  }
}
